src/test/fuzz.cpp

```diff
@@ -1,144 +1,83 @@
-// Copyright 2015 Stellar Development Foundation and contributors. Licensed
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
 // under the Apache License, Version 2.0. See the COPYING file at the root
 // of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
 
-#include "util/asio.h"
-#include "crypto/Hex.h"
+#include "test/fuzz.h"
 #include "crypto/SHA.h"
-#include "main/Application.h"
-#include "main/Config.h"
 #include "main/StellarCoreVersion.h"
-#include "overlay/OverlayManager.h"
-#include "overlay/TCPPeer.h"
-#include "overlay/test/LoopbackPeer.h"
-#include "simulation/Simulation.h"
-#include "test/test.h"
-#include "util/Fs.h"
-#include "util/Logging.h"
-#include "util/Math.h"
-#include "util/Timer.h"
+#include "test/FuzzerImpl.h"
 #include "util/XDRStream.h"
 
-#include "test/fuzz.h"
-
-#include <signal.h>
+#include <algorithm>
+#include <autocheck/autocheck.hpp>
 #include <xdrpp/autocheck.h>
-#include <xdrpp/printer.h>
-
 /**
  * This is a very simple fuzzer _stub_. It's intended to be run under an
- * external fuzzer with some fuzzing brains, AFL or fuzzgrind or whatever.
+ * external fuzzer with some fuzzing brains, at this time, preferably AFL.
  *
  * It has two modes:
  *
- *   - In gen-fuzz mode it spits out a small file containing a handful of
- *     random StellarMessages. This is the mode you use to generate seed data
- *     for the external fuzzer's corpus.
+ *   - In genfuzz mode it spits out a small file containing a handful of
+ *     random FuzzTransactionInputs or StellarMessages. This is the mode you use
+ *     to generate seed data for the external fuzzer's corpus.
  *
- *   - In fuzz mode it reads back a file and appplies it to a pair of of
- *     stellar-cores in loopback mode, cranking the I/O loop to simulate
- *     receiving the messages one by one. It exits when it's read the input.
- *     This is the mode the external fuzzer will run its mutant inputs through.
+ *   - In fuzz mode it reads back a file and appplies it to a stellar-core
+ *     instance, applying but not committing transactions one by one to simulate
+ *     certain transaction/overlay scenarios. It exits when it's applied the
+ *     input. This is the mode the external fuzzer will run its mutant inputs
+ *     through.
  *
  */
 
-namespace stellar
-{
+using namespace stellar;
 
-std::string
-msgSummary(StellarMessage const& m)
-{
-    xdr::detail::Printer p(0);
-    xdr::archive(p, m.type(), nullptr);
-    return p.buf_.str() + ":" + hexAbbrev(sha256(xdr::xdr_to_msg(m)));
-}
-
-bool
-tryRead(XDRInputFileStream& in, StellarMessage& m)
+template <typename XDR_TYPE, typename Predicate>
```

Comment:
it would be cleaner to have this functionality attached to the respective fuzz class, that way you can share the validation code between genfuzz and regular fuzzing.
And if you do it like this, you won't need templates